File: workers/provider_capacity.py
"""Shared native embedding admission, including retries and ambiguous writes."""
import math
import os
import uuid

from file_runtime import database
from index_client import EMBEDDING_MODEL
from worker_metrics import count


class ProviderDeferred(Exception):
    def __init__(self, delay, available_tokens=0):
        self.delay = max(0.1, min(float(delay), 3600))
        self.available_tokens = available_tokens
        super().__init__("shared embedding capacity unavailable")


def limits():
    requests = int(os.getenv("PUFFERFS_EMBEDDING_REQUESTS_PER_MINUTE", "1024"))
    tokens = int(os.getenv("PUFFERFS_EMBEDDING_TOKENS_PER_MINUTE", "2000000"))
    if not 1 <= requests <= 1000000 or not 32768 <= tokens <= 1000000000000:
        raise ValueError("invalid shared embedding budget")
    return requests, tokens


def reserve(org_id, tokens):
    requests, limit = limits()
    reservation = uuid.uuid4().hex
    with database() as conn:
        result = conn.execute("SELECT * FROM reserve_provider_capacity(%s,%s,%s,%s,%s,%s,90)",
            (EMBEDDING_MODEL, org_id, reservation, tokens, requests, limit)).fetchone()
    if result["delay"]:
        count("embedding_admission_deferred")
        raise ProviderDeferred(result["delay"], result["available_tokens"])
    count("embedding_provider_attempts")
    return reservation


def settle(reservation, tokens):
    # Missing or changed provider metrics keep the conservative reservation.
    if type(tokens) is not int or tokens < 0:
        return
    with database() as conn:
        conn.execute("UPDATE provider_reservations SET tokens=%s WHERE id=%s", (tokens, reservation))


def cool_down(delay):
    with database() as conn:
        conn.execute("""UPDATE provider_capacity SET blocked_until=GREATEST(blocked_until,
            clock_timestamp()+make_interval(secs=>%s)) WHERE model=%s""", (delay, EMBEDDING_MODEL))


def write_embedding(job, tp, namespace, mutation, options):
    # UTF-8 bytes conservatively reserve text tokens plus prompt overhead.
    # Successful responses reconcile to the provider's measured token count.
    rows = mutation["upsert_rows"]
    costs = [len(row["content"].encode()) + 128 for row in rows]
    tokens = sum(costs)
    while True:
        try:
            reservation = reserve(job["org_id"], tokens)
            break
        except ProviderDeferred as capacity:
            # The next batch can use remaining token capacity without waiting
            # for enough room for its original size. Every smaller attempt is
            # atomically reserved again; this hint does not grant permission.
            remaining, length = capacity.available_tokens, 0
            for cost in costs:
                if cost > remaining:
                    break
                remaining -= cost
                length += 1
            if not 0 < length < len(rows):
                raise
            rows, costs = rows[:length], costs[:length]
            tokens = sum(costs)
    mutation = {"upsert_rows": rows}
    try:
        response = tp.namespace(namespace).write(**mutation, **options)
    except Exception as error:
        if getattr(error, "status_code", None) == 429:
            count("embedding_provider_throttled")
            try:
                retry = float(error.response.headers.get("retry-after", "5"))
                delay = max(1, min(retry, 3600)) if math.isfinite(retry) else 5
            except (ValueError, AttributeError):
                delay = 5
            # Turbopuffer rejects a throttled request in full. Keep the request
            # debit, but no embedding tokens were accepted for this attempt.
            settle(reservation, 0)
            cool_down(delay)
            raise ProviderDeferred(delay) from error
        raise
    # A bookkeeping failure must not hide a confirmed write/checkpoint. Its
    # original reservation remains, so failing closed costs capacity only.
    try:
        measured = (response.model_dump().get("performance") or {}).get("embedding_tokens")
        settle(reservation, measured)
    except Exception:
        count("embedding_settlement_deferred")
    return len(rows)

File: internal/server/provider_capacity.go
package server

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"math"
	"net/http"
	"os"
	"strconv"
	"time"

	"github.com/google/uuid"
)

const embeddingModel = "qwen/qwen3-embedding-8b"

type embeddingTenantKey struct{}

type providerBudget struct {
	db       *DB
	requests int
	tokens   int64
}

type embeddingCapacityError struct{ delay float64 }

func (e *embeddingCapacityError) Error() string {
	return "shared embedding capacity unavailable; retry later"
}

func (s *Server) ConfigureEmbeddingBudgetFromEnv() error {
	requests, tokens := int64(1024), int64(2000000)
	for _, option := range []struct {
		name             string
		value            *int64
		minimum, maximum int64
	}{
		{"PUFFERFS_EMBEDDING_REQUESTS_PER_MINUTE", &requests, 1, 1000000},
		{"PUFFERFS_EMBEDDING_TOKENS_PER_MINUTE", &tokens, 32768, 1000000000000},
	} {
		if raw := os.Getenv(option.name); raw != "" {
			value, err := strconv.ParseInt(raw, 10, 64)
			if err != nil || value < option.minimum || value > option.maximum {
				return fmt.Errorf("invalid %s", option.name)
			}
			*option.value = value
		}
	}
	s.tp.embeddingBudget = &providerBudget{db: s.db, requests: int(requests), tokens: tokens}
	return nil
}

func (b *providerBudget) reserve(ctx context.Context, tokens int64) (string, error) {
	org, _ := ctx.Value(embeddingTenantKey{}).(string)
	if org == "" {
		return "", errors.New("embedding request missing tenant")
	}
	if tokens > b.tokens {
		return "", errors.New("embedding request exceeds configured token budget")
	}
	id := uuid.NewString()
	var delay float64
	var available int64
	// Queries return 429 rather than waiting in a durable queue. Their tenant
	// hint expires promptly if the caller does not retry.
	err := b.db.pool.QueryRow(ctx, `SELECT * FROM reserve_provider_capacity($1,$2,$3,$4,$5,$6,1)`,
		embeddingModel, org, id, tokens, b.requests, b.tokens).Scan(&delay, &available)
	if err != nil {
		return "", err
	}
	if delay > 0 {
		return "", &embeddingCapacityError{delay: delay}
	}
	return id, nil
}

func (b *providerBudget) settle(id string, response []byte) {
	var metadata struct {
		Performance struct {
			Tokens *int64 `json:"embedding_tokens"`
		} `json:"performance"`
	}
	if json.Unmarshal(response, &metadata) != nil || metadata.Performance.Tokens == nil || *metadata.Performance.Tokens < 0 {
		return
	}
	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
	defer cancel()
	// On bookkeeping failure, the conservative reservation survives. Do not
	// turn a confirmed provider response into a second billable attempt.
	_, _ = b.db.pool.Exec(ctx, `UPDATE provider_reservations SET tokens=$2 WHERE id=$1`, id, *metadata.Performance.Tokens)
}

func (b *providerBudget) throttled(id string, header http.Header) error {
	delay := 5.0
	if value, err := strconv.ParseFloat(header.Get("Retry-After"), 64); err == nil && !math.IsNaN(value) && !math.IsInf(value, 0) {
		delay = min(3600, max(1, value))
	}
	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
	defer cancel()
	tx, err := b.db.pool.Begin(ctx)
	if err != nil {
		return err
	}
	defer tx.Rollback(ctx)
	if _, err = tx.Exec(ctx, `UPDATE provider_capacity SET blocked_until=GREATEST(blocked_until,
		clock_timestamp()+make_interval(secs=>$2)) WHERE model=$1`, embeddingModel, delay); err != nil {
		return err
	}
	if _, err = tx.Exec(ctx, `UPDATE provider_reservations SET tokens=0 WHERE id=$1`, id); err != nil {
		return err
	}
	if err = tx.Commit(ctx); err != nil {
		return err
	}
	return &embeddingCapacityError{delay: delay}
}

func queryEmbeddingTokens(body any) int64 {
	switch value := body.(type) {
	case TPQuery:
		return rankEmbeddingTokens(value.RankBy)
	case map[string]any:
		var total int64
		if queries, ok := value["queries"].([]TPQuery); ok {
			for _, query := range queries {
				total += rankEmbeddingTokens(query.RankBy)
			}
		}
		return total
	}
	return 0
}

func rankEmbeddingTokens(rank any) int64 {
	items, ok := rank.([]any)
	if !ok || len(items) == 0 {
		return 0
	}
	name, _ := items[0].(string)
	if len(items) == 2 && name == "Embed" {
		if text, ok := items[1].(string); ok {
			return int64(len(text)) + 128
		}
	}
	var total int64
	for _, item := range items {
		total += rankEmbeddingTokens(item)
	}
	return total
}
